src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java

```diff
@@ -0,0 +1,165 @@
+                    "SELECT * FROM shopping_carts c "
+                    + "JOIN shopping_carts_products cp ON c.id = cp.cart_id "
+                    + "JOIN products p on cp.product_id = p.id "
+                    + "WHERE c.deleted = FALSE",
+                    ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_UPDATABLE);
+            ResultSet resultSet = statement.executeQuery();
+            List<ShoppingCart> carts = new ArrayList<>();
+            while (resultSet.next()) {
+                carts.add(getShoppingCartFromResultSet(resultSet));
+                resultSet.previous();
+            }
+            return carts;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can not get items from DB", e);
+        }
+    }
+
+    @Override
+    public ShoppingCart update(ShoppingCart shoppingCart) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            deleteProductsFromDB(shoppingCart.getId(), connection);
+            addProductsToDB(shoppingCart, connection);
+            PreparedStatement statement = connection.prepareStatement(
+                    "UPDATE shopping_carts SET user_id = ? "
+                    + "WHERE id = ? AND deleted = FALSE");
+            statement.setLong(1, shoppingCart.getUserId());
+            statement.setLong(2, shoppingCart.getId());
+            statement.executeUpdate();
+            return shoppingCart;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Updating shopping cart "
+                                              + shoppingCart + " failed", e);
+        }
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            deleteProductsFromDB(id, connection);
+            PreparedStatement statement = connection.prepareStatement(
+                    "UPDATE shopping_carts SET deleted = TRUE WHERE id = ?");
+            statement.setLong(1, id);
+            return statement.executeUpdate() == 1;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Deleting shopping cart with id "
+                                              + id + " failed", e);
+        }
+    }
+
+    private Optional<ShoppingCart> getShoppingCartFromStatement(PreparedStatement statement)
+            throws SQLException {
+        ResultSet resultSet = statement.executeQuery();
+        if (resultSet.next()) {
+            return Optional.of(getShoppingCartFromResultSet(resultSet));
+        }
+        return Optional.empty();
+    }
+
+    private ShoppingCart getShoppingCartFromResultSet(ResultSet resultSet) throws SQLException {
+        long id = resultSet.getLong("id");
+        long userId = resultSet.getLong("user_id");
+        List<Product> products = new ArrayList<>();
+        long productId = resultSet.getLong("product_id");
+        if (productId != 0) {
```

Comment:
what do you mean by this check? AUTO_INCREMENT in SQL makes count from 1, so this statement will always be true

Reply:
when cart is empty this field is null and getLong() returns 0